ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLog.java

```diff
@@ -388,10 +388,12 @@ public TermIndex getLastEntryTermIndex() {
     try(AutoCloseableLock writeLock = writeLock()) {
       validateLogEntry(entry);
       final LogSegment currentOpenSegment = cache.getOpenSegment();
+      // The stateMachineData will be cached inside the StateMachine itself when enable stateMachineCaching.
+      LogEntryProto toAppendEntry = stateMachineCachingEnabled? ServerProtoUtils.removeStateMachineData(entry) : entry;
       if (currentOpenSegment == null) {
         cache.addOpenSegment(entry.getIndex());
         fileLogWorker.startLogSegment(entry.getIndex());
-      } else if (isSegmentFull(currentOpenSegment, entry)) {
+      } else if (isSegmentFull(currentOpenSegment, toAppendEntry)) {
```

Comment:
For this case, it should removeStateMachineData no matter stateMachineCachingEnabled is true or false.  How about the following?
```
+++ b/ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLog.java
@@ -385,13 +385,14 @@ public class SegmentedRaftLog extends RaftLog {
     if (LOG.isTraceEnabled()) {
       LOG.trace("{}: appendEntry {}", getName(), ServerProtoUtils.toLogEntryString(entry));
     }
+    final LogEntryProto removedStateMachineData = ServerProtoUtils.removeStateMachineData(entry);
     try(AutoCloseableLock writeLock = writeLock()) {
       validateLogEntry(entry);
       final LogSegment currentOpenSegment = cache.getOpenSegment();
       if (currentOpenSegment == null) {
         cache.addOpenSegment(entry.getIndex());
         fileLogWorker.startLogSegment(entry.getIndex());
-      } else if (isSegmentFull(currentOpenSegment, entry)) {
+      } else if (isSegmentFull(currentOpenSegment, removedStateMachineData)) {
         cache.rollOpenSegment(true);
         fileLogWorker.rollLogSegment(currentOpenSegment);
         checkAndEvictCache();
@@ -414,7 +415,7 @@ public class SegmentedRaftLog extends RaftLog {
           fileLogWorker.writeLogEntry(entry).getFuture();
       if (stateMachineCachingEnabled) {
         // The stateMachineData will be cached inside the StateMachine itself.
-        cache.appendEntry(ServerProtoUtils.removeStateMachineData(entry));
+        cache.appendEntry(removedStateMachineData));
       } else {
         cache.appendEntry(entry);
       }
```

Reply:
@szetszwo The problem is cache.appendEntry(entry) will put entry with StateMachineData into LogSegment#entryCache. For example, each entry occupy 8KB after remove StateMachineData, each StateMachineData is 80M, the max segment.size.max is 8MB, when `isSegmentFull` returns true, open segment has 1000 entry,  and LogSegment#entryCache is 80M * 1000, i.e. 80GB. 

```
       if (stateMachineCachingEnabled) {
         // The stateMachineData will be cached inside the StateMachine itself.
-        cache.appendEntry(ServerProtoUtils.removeStateMachineData(entry));
+        cache.appendEntry(removedStateMachineData));
       } else {
         cache.appendEntry(entry);
       }
```

Reply:
You are right.  The simple fix won't.  I guess we need this fix https://github.com/apache/incubator-ratis/pull/317#issuecomment-737819776 ?

Reply:
@szetszwo #317 (comment) need a minor change, otherwise it maybe add more entry in the open segment if stateMachineCachingEnabled is false. In the `else` branch of `isSegmentFull` , we pass `LogSegment.Op.WRITE_SEGMENT_FILE`, which count the size of entry after remove StateMachineData, but the segment.getTotalSize() is the total size of entry without remove StateMachineData. So `segment.getTotalSize() + entrySize > segmentMaxSize` becomes (total size of entries without remove StateMachineData) + (size of entry after remove StateMachineData) > segmentMaxSize

```
@@ -431,7 +432,7 @@ public class SegmentedRaftLog extends RaftLog {
     if (segment.getTotalSize() >= segmentMaxSize) {
       return true;
     } else {
-      final long entrySize = LogSegment.getEntrySize(entry);
+      final long entrySize = LogSegment.getEntrySize(entry, LogSegment.Op.WRITE_SEGMENT_FILE);
       // if entry size is greater than the max segment size, write it directly
       // into the current segment
      return entrySize <= segmentMaxSize &&
          segment.getTotalSize() + entrySize > segmentMaxSize;
```

Reply:
I see.  It is tricky.  We actually have two sizes, file size and cache size, where
- file size always excludes state machine data, and
- cache size may or may not include state machine data depending on stateMachineCachingEnabled. 

isSegmentFull(..) also has a bug when stateMachineCachingEnabled is false.